added auto gen documentation and tutorials

weidel-p opened the pull request:
Added autogenerted documentation for Lava and adapted the "Getting started" page to display tutorials directly.

This PR requires https://github.com/lava-nc/lava/pull/318 to be merged.

bamsumit reviewed (commented):
Looks awesome Phillip. Great work.

Just one comment. It looks like something is wrong with latex renderer. The equations are not rendered somehow.
![image](https://user-images.githubusercontent.com/29907126/187774708-afeeef69-1975-45fe-bf53-3e677e7b3e02.png)

bamsumit commented:
I haven't tried to build it myself. Will try that sometime later and let you know if there are any feedback.

weidel-p commented:
In the last commit I addressed the issue with rendering the Latex equation. 
Additionally, I added the auto-generated documentation for lava-optimization.

bamsumit commented:
> In the last commit I addressed the issue with rendering the Latex equation. Additionally, I added the auto-generated documentation for lava-optimization.

Great. Looks good to me then.

mathisrichter commented on getting_started/end_to_end_tutorials.rst:
```diff
@@ -0,0 +1,20 @@
+
+These are standalone tutorials that can be followed in any order.
+
+**Coming soon:**
+
+* Liquid state machine: Another classical canonical example in neuromorphic computing, illustrating a complex, spiking and recurrent neural network for time series classification.
+
+* Learning: Many neuromorphic architectures offer local on-chip learning via bio-inspired STDP learning. This tutorial provides a simple example how such learning rules can be used.
```
I suggest taking these out until we have the content ready. This gives me "Site under construction" vibes.

weidel-p replied:
I grouped all tutorials together and removed the "coming soon" parts.

mathisrichter commented on getting_started/in_depth_tutorials.rst:
```diff
@@ -0,0 +1,25 @@
+
+These tutorials walk step-by-step through the fundamental concepts of Lava introduced in :ref:`Lava Architecture <lava architecture>`. These tutorials tend to build on each other, therefore it's best to consume them in order.
+
+**Coming shortly:**
+
+* State monitoring:
+Lava offers probes to monitor the evolution of temporal state during execution. Learn how to create probes, retrieve timeseries and how to create custom visualization of network dynamics.
```
See above

weidel-p replied:
I grouped all tutorials together and removed the "coming soon" parts.

mathisrichter commented on lava/lava.rst:
```diff
@@ -1,19 +1,32 @@
+Lava API reference
+==================
+
+This page gives an overview over all public lava modules, classes and functions. 
```
lava -> Lava

mathisrichter commented on lava/lava.rst:
```diff
@@ -1,19 +1,32 @@
+  - :ref:`Magma runtime <lava/lava.magma.runtime:lava.magma.runtime>`
+    providing a frontend for execution and control
+
+A fundamental description about Lava and its key componentes can be found in :ref:`Lava Architecture <lava architecture>`.
```
The description is not fundamental, it is the concepts that are described.

"Lava's fundamental concepts and key components are described in ..."

mathisrichter reviewed (changes requested):
So far, I only browsed the changes on Github. Will have a look at the built documentation once available.

Now I looked through the generated website:
- Under "Lava's foundational concepts", remove the colons after "1. Processes:, 2. Behavioral implementations via ProcessModels: ..."
- More white-space below figures.
- The capitalization of headings is inconsistent. Either go with all words capitalized "Getting Started With Lava", only beginning and names capitalized "Getting started with Lava" or only "helper words" (I'm sure that has some fancy name) capitalized "Getting Started with Lava".
- Under "Getting started" -> "Fundamental concepts" -> Installing Lava, below "2. Getting started", this should not be a heading: When running tests if you see ‘OSError: [Errno 24] Too many open files’
- I don't think "Installing Lava" should be under "Fundamental concepts" but rather it's own heading under "Getting started".
- I suggest to not duplicate the instruction on how to clone and setup Lava here but rather point to the README.md, where this is explained. Or pull in that information here. If we have to maintain multiple places, it will get out of sync.
- The instructions on how to set up Lava on Loihi should similarly be available in the private Lava repository and pulled in from there, possibly.
- The instructions for Lava on Loihi seem to have too little detail. Rather than referring to an old email, why not put that information in the README.md or the website? What are the "Intel external vlab Machines"? How do I log in. The instructions seem to suggest to just install the release artifact with pip. Does this not require that the Lava virtual environment is set up and activated and poetry installed?
- Under "Application examples" there is the MNIST tutorial, which is a bit of a stretch already, but okay. But the "Walk through Lava" should not be listed here. It is too basic to be called an "application".
- Under "Dynamic neural fields" there are links in markdown syntax: (e.g., using [cedar](https://cedar.ini.rub.de/) or [cosivina](https://github.com/cosivina)) that are just copied in.
- Can we get rid of the "Development roadmap" and the table under "Initial release" entirely? In its current outdated form, it does not distinguish between what is currently available and what we will develop in the future.
- The list of Project committers is out of date. jlakness-intel no longer works at NCL.
- Get rid of the extra 'Table of content' in the Developers guide.
- There is a missing white space in the "will come soon" notice for the lava-dnf API documentation: API documentation will arrive shortly. In the mean time, please reviewdocstrings in code.

GaboFGuerra commented:
Thanks Philipp, this is looking great!
 
I just noticed the module_path in sync_notebook.py does not take into account the "src/lava-nc" when looking for Lava tutorials under .venv/src/lava-nc, this is the resulting folder structure when you do poetry install under the inner-source Lava repo. 

Also had to run "pip install nbsphinx" not sure if it is already on the requirements somewhere.

GaboFGuerra commented on README.md:
````diff
@@ -6,7 +6,26 @@ They will be in `_build/html`
+
+```bash
+sphinx-apidoc -o <output_path> <path-to-src-repo>/src/lava -t _templates -H -d 10 --implicit-namespace
+```
````
Apparently we need to put something after -H, to be used as prefix on the generated files. The actual word is probably irrelevant for rendering. But the command above will fail due to this missing argument.

weidel-p replied:
I removed the -H from the command. It's really not necessary and as you said faulty in this form.

weidel-p commented:
I worked on your comments and implemented them as it was possible.

> So far, I only browsed the changes on Github. Will have a look at the built documentation once available.
> 
> Now I looked through the generated website:
> 
> * Under "Lava's foundational concepts", remove the colons after "1. Processes:, 2. Behavioral implementations via ProcessModels: ..."

done.

> * More white-space below figures.

I tried to change from image to figure, which apparently adds some white space below.

> * The capitalization of headings is inconsistent. Either go with all words capitalized "Getting Started With Lava", only beginning and names capitalized "Getting started with Lava" or only "helper words" (I'm sure that has some fancy name) capitalized "Getting Started with Lava".

Done.

> * Under "Getting started" -> "Fundamental concepts" -> Installing Lava, below "2. Getting started", this should not be a heading: When running tests if you see ‘OSError: [Errno 24] Too many open files’

Done.

> * I don't think "Installing Lava" should be under "Fundamental concepts" but rather it's own heading under "Getting started".

I flattened the hierarchy there. There is no distinguishment between in-depth and end-to-end tutorials.

> * I suggest to not duplicate the instruction on how to clone and setup Lava here but rather point to the README.md, where this is explained. Or pull in that information here. If we have to maintain multiple places, it will get out of sync.

I adapted them such that they are not getting outdated that fast anymore. But I hesitate to delete a complete tutorial.

> * The instructions on how to set up Lava on Loihi should similarly be available in the private Lava repository and pulled in from there, possibly.

The private lava repo is not covered here.

> * The instructions for Lava on Loihi seem to have too little detail. Rather than referring to an old email, why not put that information in the README.md or the website? What are the "Intel external vlab Machines"? How do I log in. The instructions seem to suggest to just install the release artifact with pip. Does this not require that the Lava virtual environment is set up and activated and poetry installed?

See above.

> * Under "Application examples" there is the MNIST tutorial, which is a bit of a stretch already, but okay. But the "Walk through Lava" should not be listed here. It is too basic to be called an "application".

See above.

> * Under "Dynamic neural fields" there are links in markdown syntax: (e.g., using [cedar](https://cedar.ini.rub.de/) or [cosivina](https://github.com/cosivina)) that are just copied in.

Done.

> * Can we get rid of the "Development roadmap" and the table under "Initial release" entirely? In its current outdated form, it does not distinguish between what is currently available and what we will develop in the future.

I need more detailed feedback on how it should look like,

> * The list of Project committers is out of date. jlakness-intel no longer works at NCL.

I removed jlakness-intel but need more feedback on the current status, maybe @mgkwill can update that list?

> * Get rid of the extra 'Table of content' in the Developers guide.

Why? 

> * There is a missing white space in the "will come soon" notice for the lava-dnf API documentation: API documentation will arrive shortly. In the mean time, please reviewdocstrings in code.

Done.
